src/extensions/agent_modules/cloud_communication/eg_mqtt_broker_client/src/adu_mosquitto_utils.c

```diff
@@ -0,0 +1,261 @@
+ * @return A string containing the correlation ID. The caller is responsible for free()'ing the returned string.
+ */
+char* GenerateCorrelationIdFromTimeWithPrefix(time_t t, const char* prefix)
+{
+    return ADUC_StringFormat("%s-%ld", prefix, t);
+}
+
+/**
+ * @brief Retrieves the correlation data from MQTT properties.
+ *
+ * This function attempts to read the correlation data from the provided MQTT properties.
+ * On success, the retrieved value is stored in the passed 'value' parameter.
+ *
+ * @param[in] props Pointer to the MQTT properties from which to retrieve the correlation data.
+ * @param[out] value Pointer to a char pointer where the retrieved correlation data string will be stored.
+ *                   The caller is responsible for freeing the allocated memory using `free`.
+ *
+ * @return `true` if correlation data was successfully retrieved; otherwise, `false`.
+ *
+ * @note If the function fails to retrieve the correlation data or if any of the input parameters is NULL,
+ *       an error message will be logged.
+ */
+bool ADU_mosquitto_get_correlation_data(const mosquitto_property* props, char** value)
+{
+    if (props == NULL || value == NULL)
+    {
+        Log_Error("Null pointer (props=%p, value=%p)", props, value);
+        return false;
+    }
+
+    *value = NULL;
+
+    // Create a copy of Correlation Data and store it in the 'value' param.
+    const mosquitto_property* p = mosquitto_property_read_string(props, MQTT_PROP_CORRELATION_DATA, value, false);
+    if (p == NULL)
+    {
+        Log_Error("Failed to read correlation data");
+        return false;
+    }
+
+    return true;
+}
+
+/**
+ * @brief Check if a specific user property exists within a property list.
+ *
+ * This function iterates through an MQTT v5 property list searching for
+ * a specific user property with the given key and value.
+ *
+ * @param[in] props Pointer to the head of the MQTT v5 property list.
+ * @param[in] key The key of the user property to search for.
+ * @param[in] value The expected value for the given key.
+ *
+ * @return True if the property list contains the user property with the specified key and value, otherwise false.
+ *
+ * @note If there are multiple properties with the same key, this function will
+ * return true upon finding the first key-value match.
+ */
+bool ADU_mosquitto_has_user_property(const mosquitto_property* props, const char* key, const char* value)
+{
+    bool found = false;
+    while (props != NULL && !found)
+    {
+        char* k;
+        char* v;
+        const mosquitto_property* p =
+            mosquitto_property_read_string_pair(props, MQTT_PROP_USER_PROPERTY, &k, &v, false);
+        if (p != NULL && k != NULL && v != NULL)
+        {
+            found = strcmp(k, key) == 0 && strcmp(v, value) == 0;
+            free(k);
+            free(v);
+            if (!found)
+            {
+                props = mosquitto_property_next(props);
+            }
+        }
```

Comment:
should free both k and v in else block too, or just always free both outside of if/else afterward